### ISSUE TITLE
Kotlin compilation errors show up as exceptions after upgrade to latest dev build

### ISSUE THREAD

**Issue by bamboo** [gradle/gradle-script-kotlin#62]
A catalog of compilation errors that are now surfacing as exceptions.

Each exception we find should be added to [JetBrains YouTrack](https://youtrack.jetbrains.com/issues) with a corresponding issue on this epic.

For example:

```
> Internal error: org.jetbrains.kotlin.codegen.CompilationException: Back-end (JVM) Internal error: Failed to generate property foo
  Cause: Error type encountered: [ERROR : Type for bar] (DeferredType).
  File being compiled and position: (1,1) in /Users/bamboo/code/gradle/gradle-script-kotlin/samples/project-properties/build.gradle.kts
  PsiElement: val foo = bar
  The root cause was thrown at: KotlinTypeMapper.java:119
        at org.jetbrains.kotlin.codegen.MemberCodegen.genFunctionOrProperty(MemberCodegen.java:205)
...
```

**Comment by cbeams** [comment]
@ligee, it would be nice to have a solution to this in time for M2, given that it's something of a regression in behavior that folks would have seen against M1.

In M1 (where we were working against the `1.1.0-dev-372` build), compiler errors were much tidier and intuitive. Here (against the `1.1.0-dev-998` build), it's quite a bit harder to read, with no column indicator where the error occured, long stack traces, etc.

**Comment by cbeams** [comment]
Of course, if it's not feasible in time for M2, just let us know.

**Comment by cbeams** [comment]
De-scoped from `0.2.0`, moved back into `Kotlin Issues`. @ligee, just let us know when you think this might be able to happen, thanks.

**Comment by bamboo** [comment]
Hey @ligee, we would like to have this fixed by the next milestone. Do you think this would be feasible?

**Comment by ligee** [comment]
If I understand it correctly, this is too general to be fixed by the next milestone. There are too many similar issues - https://youtrack.jetbrains.com/issues/KT?q=CompilationException%20%23Open%20
I can imagine that these are generally considered as errors in compiler, rather than compilation errors, and therefore exposed as exceptions.
Some particular one of course could be fixed, but you need to be more specific and even better - file issues in YouTrack. Then you can ping me here with the particular issues, and then I can check the outlook with the team.

**Comment by bamboo** [comment]
Thanks, @ligee, that makes sense, in fact I now remember you told me exactly that during the Summit. I'll turn this issue into an epic to catalog the errors as we find them and add them to YouTrack.

**Comment by bamboo** [comment]
Resolved by 56828939fd16191516c29b574bfc6bf721f5f424.
